logstash-core/lib/logstash/agent.rb

```diff
@@ -388,14 +388,20 @@ def dispatch_events(converge_results)
   def start_webserver
     options = {:http_host => @http_host, :http_ports => @http_port, :http_environment => @http_environment }
     @webserver = LogStash::WebServer.new(@logger, self, options)
-    Thread.new(@webserver) do |webserver|
+    @webserver_thread = Thread.new(@webserver) do |webserver|
       LogStash::Util.set_thread_name("Api Webserver")
       webserver.run
     end
   end
 
   def stop_webserver
-    @webserver.stop if @webserver
+    if @webserver
+      @webserver.stop
+      if @webserver_thread.join(5).nil?
+        @webserver_thread.kill
```

Comment:
Thread.kill is likely to leave the server socket and any other resources hanging around until GC runs their finalizers/cleaners. This can mean that we kill the thread but the next attempt to run the webserver will fail because the socket is still in use.

Thoughts?

Reply:
@jordansissel I was hoping the internal workings of `Webserver` would deal with this?
(if not, I think resource safety on interrupt should be fixed there?)

Reply:
@jordansissel took a quick look at the Puma webserver code ... looks safe to me resource wise :)